lib/spack/spack/spec.py

```diff
@@ -3239,8 +3236,10 @@ def _constrain_dependencies(self, other):
 
         # Handle common first-order constraints directly
         changed = False
-        for name in self.common_dependencies(other):
-            changed |= self[name].constrain(other[name], deps=False)
+        common_dependencies = {x.name for x in self.dependencies()}
+        common_dependencies &= {x.name for x in other.dependencies()}
+        for name in common_dependencies:
+            changed |= self[name].constrain(other[name], deps=True)
```

Comment:
why deps=True?

Reply:
Forgot to reply here. That's because abstract specs are not depth=1 anymore, but can be depth=2. Think for instance:
```
mpileaks ^callpath %gcc@5
mpileaks ^callpath %gcc@6
```
The `gcc` is a direct dependency of callpath. I'll add some more unit tests to make this clear.

Reply:
See c0e0ad501768e32e598a661c786247dd7ac2da05